File: examples/FSharp/FSharp.Examples/Scenarios/CustomReportingSink.fs
module CustomReportingScenario

open System
open System.Threading.Tasks
open System.Net.Http

open FSharp.Control.Tasks.V2.ContextInsensitive

open NBomber.Contracts
open NBomber.FSharp

// it's a very basic CustomReportingSink example to give you a playground for writing your own custom sink
// for production purposes use https://github.com/PragmaticFlow/NBomber.Sinks.InfluxDB

type CustomReportingSink() =
    interface IReportingSink with
        member x.SinkName = "CustomReportingSink"
        member x.Init(logger, infraConfig) = ()
        member x.StartTest(testInfo: TestInfo) = Task.CompletedTask
        member x.SaveRealtimeStats(stats:NodeStats[]) = Task.CompletedTask
        member x.SaveFinalStats(stats:NodeStats[], reportFiles:ReportFile[]) = Task.CompletedTask
        member x.StopTest() = Task.CompletedTask
        member x.Dispose() = ()

let run () =

    let httpClient = new HttpClient()

    let step = Step.create("GET html", fun context -> task {
        use! response = httpClient.GetAsync("https://nbomber.com",
                                            context.CancellationToken)

        match response.IsSuccessStatusCode with
        | true  -> let size = int response.Content.Headers.ContentLength.Value
                   return Response.Ok(sizeBytes = size)
        | false -> return Response.Fail()
    })

    let scenario = Scenario.create "test_nbomber" [step]
                   |> Scenario.withWarmUpDuration(TimeSpan.FromSeconds 30.0)
                   |> Scenario.withLoadSimulations [
                       KeepConcurrentScenarios(200, TimeSpan.FromSeconds 60.0)
                   ]

    use customReportingSink = new CustomReportingSink()
    let sendStatsInterval = TimeSpan.FromSeconds 30.0

    NBomberRunner.registerScenarios [scenario]
    |> NBomberRunner.withReportingSinks([customReportingSink], sendStatsInterval)
    |> NBomberRunner.runInConsole

File: examples/Demo/Features/RealtimeReporting/CustomReportingSink/CustomReportingSink.cs
using Microsoft.Extensions.Configuration;
using NBomber.Contracts;
using NBomber.Contracts.Stats;
using NBomber.CSharp;
using Serilog;

namespace Demo.Features.RealtimeReporting.CustomReportingSink;

class CustomReportingSink : IReportingSink
{
    private ILogger _log;
    public string SinkName => nameof(CustomReportingSink);

    public Task Init(IBaseContext context, IConfiguration infraConfig)
    {
        _log = context.Logger.ForContext<CustomReportingSink>();
        return Task.CompletedTask;
    }

    public Task Start(SessionStartInfo sessionInfo) => Task.CompletedTask;
    public Task SaveRealtimeStats(ScenarioStats[] stats) => Task.CompletedTask;
    public Task SaveFinalStats(NodeStats stats) => Task.CompletedTask;
    public Task Stop() => Task.CompletedTask;

    public void Dispose()
    { }
}

public class CustomReportingExample
{
    public void Run()
    {
        var scenario = Scenario.Create("simple_scenario", async context =>
        {
            await Task.Delay(TimeSpan.FromMilliseconds(500));
            return Response.Ok(sizeBytes: 100);
        })
        .WithoutWarmUp()
        .WithLoadSimulations(Simulation.KeepConstant(1, TimeSpan.FromMinutes(1)));

        NBomberRunner
            .RegisterScenarios(scenario)
            .WithReportingSinks(new CustomReportingSink())
            .WithReportingInterval(TimeSpan.FromSeconds(10))
            .Run();
    }
}
